WifiQueue.class.nut

```
@@ -1,170 +1,206 @@
+    _connecting = false;
+    // Whether we should try networks that are not visible (because they may be hidden)
+    _hidden = null;
+    // Whether we should try open wifi networks (that are not in the list we were given)
+    _open = null;
+    // Whether we have scanned for open wifi networks to initiate warDriving yet
+    _warDriving = false;
+
+    // Callbacks
+    _onFail = null;
+    _onConnect = null;
+    _onDisconnect = null;
+
+    // -----------------------
+    // `cm` is a connection manager instance
+    // `wifis` is an [ { ssid, pw } ]
+    // `flags` is a bit-array of boolean options`
+    // `logger` is a logger object with `.log()` and `.error()` methods
+    constructor(cm, flags = null, logger = server) {
+        _cm = cm;
+
+        // Process options
+        flags = flags || 0;
+        _hidden = flags & WIFIQUEUE_HIDDEN;
+        _open   = flags & WIFIQUEUE_OPEN;
+
+        // Set ConnectionManager callbacks
+        _cm.onConnect(_didConnect.bindenv(this));
+        _cm.onTimeout(_didTimeout.bindenv(this));
+        _cm.onDisconnect(_didDisconnect.bindenv(this));
+    }
+
+    // -----------------------
+    // Attempt to connect to list of wifis
+    //
+    // Other methods of this class which call this method "recursively"
+    // e.g. `.didTimeout()` will call method with `_reset == false` to
+    // avoid resetting the wifi list each time.  When called from outside a
+    // list of wifi networks should be passed in to initialize this
+    // connecting attempt.
+    function connect(wifis = null, _reset = true) {
+        if (_reset) {
+            // Cancel any in-progress wardriving
+            _warDriving = false;
+
+             // Copy the array because we will be mutating it, and we need to be the
+            // only ones
+            _wifis = clone(wifis || []);
+
+            // Remove invalid entries
+            for (local i = _wifis.len() - 1; i >= 0; i--) {
```

I probably wouldn't check if they're valid. Would just assume they are and let it fail if they're not.

---

"let it fail" could mean an unhandled exception which causes the device to reboot and then repeatedly fail to connect